File: answers/leetcode/Longest Valid Parentheses/Longest Valid Parentheses.py
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """

        longest = 0
        stack = []
        begin = -1
        for i in range(len(s)):
            if '(' == s[i]:
                stack.append(i)
            elif ')' == s[i]:
                if 0 == len(stack):
                    begin = i
                else:
                    stack.pop()
                    if 0 == len(stack):
                        length = i - begin
                    else:
                        length = i - stack[-1]
                    longest = max(longest, length)
        return longest

File: answers/leetcode/Longest Valid Parentheses/Longest Valid Parentheses.cpp
//@result 229 / 229 test cases passed. Status: Accepted Runtime: 12 ms Submitted: 0 minutes ago You are here! Your runtime beats 11.28% of cpp submissions.

class Solution {
public:
	int longestValidParentheses(string s) {
		size_t len_max = 0;
		stack<size_t> sta;
		size_t p = 0;
		while (s[p]) {
			if ('(' == s[p]) {
				sta.push(p);
			}
			else {
				if (! sta.empty() && '(' == s[sta.top()]) {
					sta.pop();
					len_max = max(len_max, p - (sta.empty() ? -1 : sta.top()));
				}
				else {
					sta.push(p);
				}
			}
			++ p;
		}
		return len_max;
	}
};
